### ISSUE TITLE
Deletion occcurence

### ISSUE THREAD

**Issue by SSwethaSel0609** [k2-fsa/icefall#1373]
I'm getting 77% wer. 
%WER 77.58% [21709 / 27984, 465 ins, 13810 del, 7434 sub ]
how to reduce the wer and what will be the reason for getting high wer

**Comment by marcoyang1998** [comment]
Could you please share some decoding examples? You can find them in the decoding output files.

**Comment by SSwethaSel0609** [comment]
Errors: 465 insertions, 13810 deletions, 7434 substitutions, over 27984 reference words (6740 correct)
Search below for sections starting with PER-UTT DETAILS:, SUBSTITUTIONS:, DELETIONS:, INSERTIONS:, PER-WORD STATS:

PER-UTT DETAILS: corr or (ref->hyp)
002_Kannur_F_18_monologue_00293:        (കലാമണ്ഡലം കല്യാണികുട്ടിയമ്മ കേരളത്തിലെ പ്രീശസ്തയായ മോഹിനിയാട്ട നർത്തകിയാആയിരുന്നു->കലാമടലം കല്യാണിക്കുട്ടിയമ്മ കേരളത്തിലെത്തി)_00924:        എങ്കിലും (ഇൻട്രാനെറ്റ്വെയർ->ഇൻട്രാനറ്റ്) എന്ന (പേര് നെറ്റെവെയറിൻ്റെ അഞ്ചാംപതിപ്പിൽ ഉപേക്ഷിച്ചു->പേരിൽ നെറ്റ്)ur_F_18_monologue_01144:        ഇപ്പോൾ (സ്റ്റീൽ കളിമണ്ണ് പ്ളാസ്റ്റിക് ഗ്ലാസ്സ് തുടങ്ങിയ സാധനങ്ങൾകൊണ്ട് തട്ടങ്ങൾ ഉണ്ടാക്കുക പതിവായിരിക്കുന്നുറ്റേറ്റ് ക്ലാസ് തുടങ്ങിയവയാണ്)
002_Kannur_F_18_monologue_01228:        നേരത്തെ നന്നായി (കഴുകി ഉണക്കി ബ്രൌൺ പേപ്പറിൽ പൊതിഞ്ഞാണ് ചൂടാക്കാൻ വയ്ക്കുന്നത്->കഴിഞ്ഞുണയ്ക്കയും പ്രവര്ത്തി)_Kannur_F_18_monologue_01317:        ഇവിടെ ബുദ്ധമതത്തിന്റെ (തെളിവുകളും അവശേഷിക്കുന്നു->*)
003_Kannur_M_26_monologue_01607:        ഈ അവസ്ഥയിൽ (ഉളള ഒരു ഗ്രഹം->രൂ)
003_Kannur_M_26_monologue_01641:        (ചിവാർളി->ചിവാർലി) പ്രണയകാവ്യമായ (ഒർലാണ്ടോ->ഒർലാൻഡോ) ഇന്നാമോർട്ടോ (എന്നിവയുടെ പേരിലാണ് ഇദ്ദ->*)പ്രധാനമായി ഓർമിക്കപ്പെടുന്നത്
003_Kannur_M_26_monologue_01746:        (ബിഷപ്പ് ജെറോമിന്റെ ഭൗതികാവശിഷ്ടങ്ങൾ അടക്കം ചെയ്തതു ഇവിടെയാണ്->ബിഷബ് ജറോമിന്റെ ഭൗതികാവശ്യമായത്)
003_Kannur_M_26_monologue_01861:        ഈ (സംഭവത്തിന് ശേഷം കുമാരപ്പണിക്കരടക്കമുള്ള സമര നേതാക്കൾക്ക് ഒളിവിൽ പോകേണ്ടതായി വന്നു->സംഭവത്തിനുശേത്) കുമാരപ്പണിക്കരടക്കുമുള്ള
007_Kannur_M_21_monologue_00975:        (അപർണ ബാലൻ കോഴിക്കോദുനിന്നുള്ള ബാഡ്മിന്റൺ കാളിക്കാരിയാണു->അപർണബാലൻ കോഴിക്കോതു നിന്നുള്ള ആദ്യം)
007_Kannur_M_21_monologue_02076:        മലബാർ (ജൂതന്മാരാണ് ഇത് നിർമ്മിച്ചത്->ജൂതരി)

**Comment by marcoyang1998** [comment]
Thanks. Could you please also show the loss curves from tensorboard logs?

**Comment by SSwethaSel0609** [comment]
![Screenshot (54)](https://github.com/k2-fsa/icefall/assets/146105702/1854322d-8a9c-4747-94ff-609838ea0a0e)
![Screenshot (55)](https://github.com/k2-fsa/icefall/assets/146105702/71d4e7d6-f07d-4bf2-9745-799770ca5f8c)
![Screenshot (56)](https://github.com/k2-fsa/icefall/assets/146105702/5727bfd7-6145-473a-9785-855a3d762079)
![Screenshot (57)](https://github.com/k2-fsa/icefall/assets/146105702/c0c587c4-91a3-4a47-b1d9-0eb9e288c9f0)
![Screenshot (58)](https://github.com/k2-fsa/icefall/assets/146105702/82509c32-1ced-4c81-a11c-0115be69da19)
![Screenshot (59)](https://github.com/k2-fsa/icefall/assets/146105702/cb53a509-cb19-4e98-91dd-c5fca9522da3)

**Comment by marcoyang1998** [comment]
The pruned_loss looks quite high. Your model seems underfitting. How much data are you using for training?

**Comment by SSwethaSel0609** [comment]
train -215 hrs of data

**Comment by desh2608** [comment]
* What recipe are you using?
* How did you prepare the data?

Some remarks:
* Make sure you have pre-shuffled the whole manifest (see [here](https://github.com/k2-fsa/icefall/blob/231bbcd2b638826a94cf019fa31ae8683d3552ee/egs/spgispeech/ASR/prepare.sh#L111)).
* For a well converged model, pruned loss should ideally be ~0.1 or lower. Check the training logs for TED-LIUM here (which is ~400h so similar to your data size): https://tensorboard.dev/experiment/AKXbJha0S9aXyfmuvG4h5A/#scalars

**Comment by SSwethaSel0609** [comment]
I'm using the below script


#!/bin/bash

set -eou pipefail

if grep -q docker /proc/1/cgroup; then 
   scripts=/scripts
   local=/local
else
   scripts=./scripts
   local=./local
fi

config_path="${scripts}/config.json"

cuda_visible_devices=$(cat $config_path | jq  ."cuda_visible_devices")
export CUDA_VISIBLE_DEVICES=$cuda_visible_devices
# ------------------------- Begin Configuration -------------------------
language=$(cat $config_path | jq -r ."language")

kaldi_train_dir=$(cat $config_path | jq  ."kaldi_train_dir"|sed 's/"//g')
kaldi_val_dir=$(cat $config_path | jq  ."kaldi_val_dir"|sed 's/"//g')
kaldi_test_dirs=$(cat $config_path | jq  ."kaldi_test_dirs"|sed 's/"//g')
musan_download_dir=$(cat $config_path | jq  ."musan_download_dir"|sed 's/"//g')

on_the_fly_feats=$(cat $config_path | jq -r ."on_the_fly_feats")
bpe_vocab_size=$(cat $config_path | jq -r ."bpe_vocab_size")
sampling_rate=$(cat $config_path | jq -r ."sampling_rate")
feature_dim=$(cat $config_path | jq -r ."feature_dim")
num_jobs=$(cat $config_path | jq -r ."num_jobs")
prepare_dev_from_train=$(cat $config_path | jq -r ."prepare_dev_from_train")
dev_from_train_percent=$(cat $config_path | jq -r ."dev_from_train_percent")

lang_dir=data/${language}/lang_bpe_${bpe_vocab_size}
manifest_dir="data/${language}/manifests"
cut_dir="data/${language}/manifests"
lm_dir="data/${language}/lm"

stage=$(cat $config_path | jq -r ."stage")
stop_stage=$(cat $config_path | jq -r ."stop_stage")
kaldi_root_dir=$(cat $config_path | jq -r ."kaldi_root_dir")
icefall_root_dir=$(cat $config_path | jq -r ."icefall_root_dir")
# ------------------------- End Configuration -------------------------

shared=$icefall_root_dir/icefall/shared
export PYTHONPATH=$icefall_root_dir

log() {
  # This function is from espnet
  local fname=${BASH_SOURCE[1]##*/}
  echo -e "$(date '+%Y-%m-%d %H:%M:%S') (${fname}:${BASH_LINENO[0]}:${FUNCNAME[1]}) $*"
}

if [ $stage -le 0 ] && [ $stop_stage -ge 0 ]; then
    if [ ! -d $musan_download_dir/musan ]; then
        log "Stage 0: Downloading MUSAN data."
        lhotse download musan $musan_download_dir
    fi
fi

if [ $stage -le 1 ] && [ $stop_stage -ge 1 ]; then
    log "Stage 1: Preparing icefall manifests."
    mkdir -p $manifest_dir
    for data in $kaldi_train_dir; do
        bdata=$(basename $data)
        log "Processing $bdata and $data "
        python3 $scripts/data_parser_kaldi.py \
            -i $data \
            -o $manifest_dir/$bdata \
            -s $sampling_rate \
            -k $kaldi_root_dir
    done

    log "Processing musan"
    # We assume that you have downloaded the musan corpus
    # to data/musan
    if [ ! -e $manifest_dir/.musan.done ]; then
        lhotse prepare musan $musan_download_dir/musan $manifest_dir/musan
    fi
fi

if [ $stage -le 2 ] && [ $stop_stage -ge 2 ]; then
    log "Stage 2: Preparing test cuts."
    for data in $kaldi_test_dirs $kaldi_val_dir; do
        bdata=$(basename $data)
        log "Processing $bdata"
        python3 $scripts/prepare_cuts.py \
            -c "${language}_${bdata}_cuts" \
            -d $manifest_dir/$bdata \
            -s $sampling_rate \
            -o $cut_dir \
            -v
        log "Succesfully stored $cut_dir/${language}_${bdata}_cuts.jsonl.gz"
    done
fi

if [ $stage -le 3 ] && [ $stop_stage -ge 3 ]; then
    log "Stage 3: Prepare musan cuts."
    $local/prepare_musan_cuts.py \
        -m $manifest_dir/musan \
        -o $cut_dir \
        -s $sampling_rate
fi

if [ $stage -le 4 ] && [ $stop_stage -ge 4 ]; then
    log "Stage 4: Preparing training cuts."
    data=$kaldi_train_dir
    bdata=$(basename $data)
    python3 $scripts/prepare_cuts.py \
            -c "${language}_${bdata}_cuts" \
            -d $manifest_dir/$bdata \
            -s $sampling_rate \
            -o $cut_dir \
            -p \
            -v
    log "Succesfully stored $cut_dir/${language}_${bdata}_cuts.jsonl.gz"
fi

if [ $stage -le 5 ] && [ $stop_stage -ge 5 ]; then
    log "Stage 5: Dump transcripts from training and validation sets for LM"
    mkdir -p $lm_dir
    
    dev_args=""
    [ ! -z $kaldi_val_dir ] && dev_args="$kaldi_val_dir/text"
    #echo "$dev_args"

    cut -d' ' -sf2- $kaldi_train_dir/text $dev_args \
         > $lm_dir/transcript_words.txt
        
    tr ' ' '\n' < $lm_dir/transcript_words.txt \
        | sed -r "/^\s*$/d" \
        | sort -u > $lm_dir/words_full.txt
    #echo "starting the special remover"
    # Separate regular and special words. The special words are enclosed in
    # <> or []. E.g. <filler>, [unintelligible/], etc.
    (echo "!SIL"; echo "<unk>"; (
        grep '<.*>\|\[.*\]' $lm_dir/words_full.txt | grep -v "<unk>\|<eps>")) \
        > $lm_dir/words_special.txt
     echo "None"
     grep -v '<.*>\|\[.*\]' $lm_dir/words_full.txt | grep -v "!SIL" \
        > $lm_dir/words.txt
     echo "before words.txt"
     grep -v '<.*>\|\[.*\]' $lm_dir/words_full.txt > $lm_dir/words.txt
fi

if [ $stage -le 6 ] && [ $stop_stage -ge 6 ]; then
    log "Stage 6: Prepare BPE based lang"
    mkdir -p $lang_dir
    echo $local
    # Prepare words.txt
    #(echo "<eps>"; cat $lm_dir/words_special.txt $lm_dir/words.txt; \
     #   echo "#0"; echo "<s>"; echo "</s>") | awk '{print $0, NR-1}' \
      #  >$lang_dir/words.txt

    if [ ! -f $lang_dir/bpe.model ]; then
        $local/train_bpe_model.py \
        --lang-dir $lang_dir \
        --vocab-size $bpe_vocab_size \
        --transcript $lm_dir/transcript_words.txt \
        --special-words $lm_dir/words_special.txt
    fi

    if [ ! -f $lang_dir/L_disambig.pt ]; then
        # TODO: This script does not add entries for special words.
        # So they are treated as <unk> by default. Change this behaviour.
        #$local/prepare_lang_bpe.py --lang-dir $lang_dir \
        $local/prepare_lang_bpe.py  --lang-dir $lang_dir \
            --special-words $lm_dir/words_special.txt

        log "Validating $lang_dir/lexicon.txt"
        $local/validate_bpe_lexicon.py \
        --lexicon $lang_dir/lexicon.txt \
        --bpe-model $lang_dir/bpe.model
    fi
fi

if [ $stage -le 7 ] && [ $stop_stage -ge 7 ]; then
    log "Stage 7: Prepare arpa file"
    $shared/make_kn_lm.py \
        -ngram-order 3 \
        -text $lm_dir/transcript_words.txt \
        -lm $lm_dir/G_3_gram.arpa
    # lmplz \
    #     --text $lang_dir/transcript_words.txt \
    #     -o 3 \
    #     --arpa $lm_dir/G_3_gram.arpa

fi

if [ $stage -le 8 ] && [ $stop_stage -ge 8 ]; then
    log "Stage 8: Prepare G"
    # We assume you have install kaldilm, if not, please install
    # it using: pip install kaldilm

    mkdir -p $lm_dir
    if [ ! -f $lm_dir/G_3_gram.fst.txt ]; then
        # It is used in building HLG
        python3 -m kaldilm \
        --read-symbol-table="$lang_dir/words.txt" \
        --disambig-symbol='#0' \
        --max-order=3 \
        $lm_dir/G_3_gram.arpa > $lm_dir/G_3_gram.fst.txt
    fi
fi

# Compile LG for RNN-T fast_beam_search decoding
if [ $stage -le 9 ] && [ $stop_stage -ge 9 ]; then
    log "Stage 9: Compile LG"
    #$local/compile_lg.py --lang-dir $lang_dir
    $local/compile_lg.py --lang-dir $lang_dir --lm-dir $lm_dir
fi

# Extract features
if [ $stage -le 10 ] && [ $stop_stage -ge 10 ]; then
    log "Stage 10: Extract features"
    
    if $on_the_fly_feats; then
        log "Skipping since on the fly features are enabled"
    else
        log "Processing musan"
        $scripts/compute_store_fbank_cuts.py \
            -i "$cut_dir/musan_cuts.jsonl.gz" \
            -o "$cut_dir/.temp_new_musan_cuts.jsonl.gz" \
            -f "$cut_dir/musan_features" \
            -n $feature_dim \
            -j $num_jobs

            mv $cut_dir/.temp_new_musan_cuts.jsonl.gz $cut_dir/musan_cuts.jsonl.gz
        log "Successfully updated musan cuts"

        for data in $kaldi_test_dirs $kaldi_val_dir $kaldi_train_dir; do
            bdata=$(basename $data)
            log "Processing $bdata"
            $scripts/compute_store_fbank_cuts.py \
                -i $cut_dir/${language}_${bdata}_cuts.jsonl.gz \
                -o $cut_dir/.temp_new_${language}_${bdata}_cuts.jsonl.gz \
                -f $cut_dir/${language}_${bdata}_features \
                -n $feature_dim \
                -j $num_jobs
            
            mv $cut_dir/.temp_new_${language}_${bdata}_cuts.jsonl.gz $cut_dir/${language}_${bdata}_cuts.jsonl.gz
            log "Succesfully updated $bdata cuts"
        done
    fi
fi

if [ $stage -le 11 ] && [ $stop_stage -ge 11 ]; then
    log "Stage 11: Extracting dev set from train set, if specified "
# when Kaldi dev directory is empty, we will take out part of train set and prepare dev set
    if [ -z $kaldi_val_dir ] && [ $prepare_dev_from_train == "true" ]; then 
        log "Building dev set from train set (Splitting train set into train & dev set)"
        train_bdata=$(basename $kaldi_train_dir)
        full_train_cutset=$cut_dir/${language}_${train_bdata}_cuts.jsonl.gz
        train_cutset=$cut_dir/${language}_train_${train_bdata}_cuts.jsonl.gz
        dev_cutset=$cut_dir/${language}_dev_${train_bdata}_cuts.jsonl.gz

        full_tr_num_lines=`zcat ${full_train_cutset} | wc -l` 
        dev_file_lines=`echo "($full_tr_num_lines * $dev_from_train_percent)/100" | bc`
        let tr_file_lines=$full_tr_num_lines-$dev_file_lines

        log "Count of cuts before splitting : Train set = $tr_file_lines , Dev set = $dev_file_lines , & Full train set (before splitting) = $full_tr_num_lines"
        zcat ${full_train_cutset} | shuf > .shuflled_full_train_cutset
        
        split -l "$tr_file_lines" -a 1 -d .shuflled_full_train_cutset .xzl_temp

        gzip .xzl_temp0 .xzl_temp1
        mv .xzl_temp0.gz ${train_cutset}
        mv .xzl_temp1.gz ${dev_cutset}

         log "Count of cuts after splitting  : Train set = `zcat ${train_cutset} | wc -l` , Dev set = `zcat ${dev_cutset} | wc -l` "
        log "stored train at ${train_cutset} "
        log "stored dev at ${dev_cutset}"
        new_full_train=`dirname ${full_train_cutset}`/full_`basename ${full_train_cutset}`
        log "Moving ${full_train_cutset} --> ${new_full_train}"

        mv ${full_train_cutset}  ${new_full_train}

        rm .shuflled_full_train_cutset 
    fi
fi

**Comment by SSwethaSel0609** [comment]
I have used the below script to prepare the data
[prepare.txt](https://github.com/k2-fsa/icefall/files/13276215/prepare.txt)
